File: core/utils/collision.js
// import { ctx } from "../core/canvas.js";

function circleCollision(a, b) {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  const dist = dx * dx + dy * dy;
  const sizeSq = a.r * a.r + b.r * b.r

  return dist < sizeSq;
}

function boxCollision(a, b) {
  return (
    a.x < b.x + b.w &&
    a.x + a.w > b.x &&
    a.y < b.y + b.h &&
    a.y + a.h > b.y
  );
}

export function boxCircleCollision(box, circle) {
  const closestX = Math.max(box.x, Math.min(circle.x, box.x + box.w));
  const closestY = Math.max(box.y, Math.min(circle.y, box.y + box.h));

  // ctx.fillStyle = "gold";
  // ctx.beginPath();
  // ctx.arc(closestX, closestY, 5, 0, Math.PI * 2)
  // ctx.fill();

  const dx = circle.x - closestX;
  const dy = circle.y - closestY;

  return (dx * dx + dy * dy) < (circle.r * circle.r);
}

export function checkCollision(a, b) {
  if (a.shape === "box" && b.shape === "box") {
    return boxCollision(a, b);
  }

  if (a.shape === "circle" && b.shape === "circle") {
    return circleCollision(a, b);
  }

  if (a.shape === "box" && b.shape === "circle") {
    return boxCircleCollision(a, b);
  }

  if (a.shape === "circle" && b.shape === "box") {
    return boxCircleCollision(b, a);
  }

  return false;
}


function projectPolygon(axis, vertices) {
  const projections = vertices.map(v => (v.x * axis.x + v.y * axis.y));
  return {
    min: Math.min(...projections),
    max: Math.max(...projections)
  };
}

function isSeparatingAxis(axis, vertices1, vertices2) {
  const proj1 = projectPolygon(axis, vertices1);
  const proj2 = projectPolygon(axis, vertices2);

  return proj1.max < proj2.min || proj2.max < proj1.min;
}

function getEdges(vertices) {
  let edges = [];
  for (let i = 0; i < vertices.length; i++) {
    const next = (i + 1) % vertices.length;
    edges.push({
      x: vertices[next].x - vertices[i].x,
      y: vertices[next].y - vertices[i].y
    });
  }
  return edges;
}


function project(vertices, axis) {
  let min = Infinity;
  let max = -Infinity;

  for (const vertex of vertices) {
    const projection = vertex.x * axis.x + vertex.y * axis.y;

    if (projection < min) min = projection;
    if (projection > max) max = projection;
  }

  return { min, max };
}

function getAxes(vertices) {
  const axes = [];
  for (let i = 0; i < vertices.length; i++) {
    const next = (i + 1) % vertices.length;
    const edge = {
      x: vertices[next].x - vertices[i].x,
      y: vertices[next].y - vertices[i].y
    };

    const length = Math.hypot(edge.x, edge.y);

    const normal = { x: -edge.y / length, y: edge.x / length };
    axes.push(normal);
  }
  return axes;
}

export function isSeperatingAxes(poly1, poly2) {
  let smallestOverlap = Infinity;
  let smallestAxis = null;

  const axes1 = getAxes(poly1);
  const axes2 = getAxes(poly2);

  for (const axis of [...axes1, ...axes2]) {
    const proj1 = project(poly1, axis);
    const proj2 = project(poly2, axis);
    if (proj1.max < proj2.min || proj2.max < proj1.min) {
      return {
        collision: false
      };
    }

    const overlap = Math.min(proj1.max, proj2.max) - Math.max(proj1.min, proj2.min);

    if (overlap < smallestOverlap) {
      smallestOverlap = overlap;
      smallestAxis = axis;
    }
  }

  return {
    collision: true,
    smallestOverlap,
    smallestAxis
  };
}

File: void_front/game/core/utils/collision.py
from __future__ import annotations

import numpy as np


def boxCircleCollision(box, circle):
    closestX = max(box.x, min(circle.x, box.x + box.w))
    closestY = max(box.y, min(circle.y, box.y + box.h))
    dx = circle.x - closestX
    dy = circle.y - closestY
    return (dx * dx + dy * dy) < (circle.r * circle.r)


def _circle_collision(a, b):
    dx = a.x - b.x
    dy = a.y - b.y
    dist = dx * dx + dy * dy
    sizeSq = a.r * a.r + b.r * b.r
    return dist < sizeSq


def _box_collision(a, b):
    return a.x < b.x + b.w and a.x + a.w > b.x and a.y < b.y + b.h and a.y + a.h > b.y


def checkCollision(a, b):
    if a.shape == "box" and b.shape == "box":
        return _box_collision(a, b)
    if a.shape == "circle" and b.shape == "circle":
        return _circle_collision(a, b)
    if a.shape == "box" and b.shape == "circle":
        return boxCircleCollision(a, b)
    if a.shape == "circle" and b.shape == "box":
        return boxCircleCollision(b, a)
    return False


def _project(vertices, axis):
    projections = np.dot(vertices, axis)
    return float(np.min(projections)), float(np.max(projections))


def _get_axes(vertices):
    axes = []
    for i in range(len(vertices)):
        next_i = (i + 1) % len(vertices)
        edge = vertices[next_i] - vertices[i]
        length = np.hypot(edge[0], edge[1])
        if length == 0:
            continue
        axes.append(np.array([-edge[1] / length, edge[0] / length], dtype=float))
    return axes


def isSeperatingAxes(poly1, poly2):
    vertices1 = np.array([[v["x"], v["y"]] if isinstance(v, dict) else [v.x, v.y] for v in poly1], dtype=float)
    vertices2 = np.array([[v["x"], v["y"]] if isinstance(v, dict) else [v.x, v.y] for v in poly2], dtype=float)
    smallestOverlap = float("inf")
    smallestAxis = None
    axes = _get_axes(vertices1) + _get_axes(vertices2)

    for axis in axes:
        min1, max1 = _project(vertices1, axis)
        min2, max2 = _project(vertices2, axis)
        if max1 < min2 or max2 < min1:
            return {"collision": False}
        overlap = min(max1, max2) - max(min1, min2)
        if overlap < smallestOverlap:
            smallestOverlap = overlap
            smallestAxis = {"x": float(axis[0]), "y": float(axis[1])}

    return {"collision": True, "smallestOverlap": smallestOverlap, "smallestAxis": smallestAxis}
